### DISCUSSION TITLE
CPU usage high when playing MP3 from LittleFS

### DISCUSSION THREAD

**Discussion by ld21** [pschatzmann/arduino-audio-tools#1006] [Q&A] [answered]
Hi Phil and everyone

I'm playing MP3 files from LittleFS well on a ESP32 Lolin D32 over I2S using your libraries – thank you (see attachement). However, if I try to use MP3-files as sound effects in a game that I have programmed, it seems as if there is not enough CPU power. Both visuals and audio lag. I was wondering, if WAV-files would be less demanding because they are not encoded. Or would it be be possible to use one core for audio and another for the rest? Unfortunately, I wasn't able to find a WAV to I2S example, so I could not try it out. What do you think of the idea of using WAV-files? Is there an example of WAV to I2S? Thank you for pointing it out.

Regards, Dani

[player-littlefs-i2s_geht_gut_im_8x8_handheld.ino.zip](https://github.com/pschatzmann/arduino-audio-tools/files/12792186/player-littlefs-i2s_geht_gut_im_8x8_handheld.ino.zip)

**Comment by pschatzmann** [comment, marked as answer]
I don't understand why you would need a separate example for this: You would just use a different decoder!
I guess it might be a good idea to use a separate core for the audio decoding of MP3: see [here](https://github.com/pschatzmann/arduino-audio-tools/wiki/Multicore-Processing)

PCM WAV is much more efficient in the sense that it does not need any decoding, but we can just copy the data as is. If you are looking for an efficient compressed audio format ADPCM would be a perfect choice (e.g. in combination with WAV).

**Reply by ld21** [reply]
Ok, I went wrong from an old example I found somewhere in your libraries ... Thank you for your hint: I now chose the correct WAVDecoder and the code works for playing WAV-files (see below). However, I get clicking sounds during the playing and at the end – even if only the sketch is running (listen [here](https://wunderwald.ch/pics/clicks.mp4)). Is there something I might consider to eliminate that?

Thank you, Dani

`/**
 * @file player-littlefs-i2s.ino
 * @brief example using the LittleFS library
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#include "AudioTools.h"
#include "AudioLibs/AudioSourceLittleFS.h"
#include "AudioCodecs/CodecWAV.h"

const char *startFilePath="/";
const char* ext="wav";
AudioSourceLittleFS source(startFilePath, ext);
I2SStream i2s;
WAVDecoder decoder;
AudioPlayer player(source, i2s, decoder);

void setup() {
  Serial.begin(57600);
  pinMode(4, INPUT_PULLUP);
  pinMode(16, INPUT_PULLUP);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  // begin processing
  auto cfg = i2s.defaultConfig();
  // Standard-I2S-Pins und vieles Anderes stehen unter:
  // https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioConfig.h
  cfg.pin_ws = 15;
  cfg.pin_bck = 14;
  cfg.pin_data = 25;
  // _____________________
  cfg.sample_rate = 8000;
  cfg.channels = 1;
  i2s.begin(cfg);
}

void loop() {
  player.copy();
  if(digitalRead(4) == LOW){
    if(player.isActive()){player.end();}
    source.setFileFilter("treffer.wav");
    player.begin();
  }
  if(digitalRead(16) == LOW){
    if(player.isActive()){player.end();}
    source.setFileFilter("schuss.wav");
    player.begin();
  }
}`

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/wiki/Introduction#logging
